### PR TITLE
 #164069188 Authenticated users should tag articles

### PR DESCRIPTION
**Pull request by CryceTruly**
**What does this PR do**
This PR enables authenticated users to tag their articles

**Description of tasks to be completed**
- Add tags to articles

**How can this be tested??**
- `clone the repo`
- `checkout ft-tag-articles-164069188`
- `npm install && npm start`
- `log in with your account credentials`
- `Create an article and add some tags`

**Any background context you want to provide**
  > Tagging refers to the metadata that is assigned to a piece of content by the content creator and the readers/users of the content. Tagging allows for organization and or categorization of content

#### Applicable screenshots
At the bottom of create and edit article page,a user can add a tag to the corresponding article

Screenshot 2019-04-19 at 23 56 59
Screenshot 2019-04-19 at 23 56 31

The application also lists all tags.This will be helpfull when displaying articles with similar tags
Screenshot 2019-04-23 at 14 10 51

**What are the relevant pivotal tracker stories**
- [#164069188 ](https://www.pivotaltracker.com/story/show/164069188 )

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -65,6 +65,7 @@
         "jwt-decode": "^2.2.0",
         "prop-types": "^15.7.2",
         "react": "^16.8.6",
+        "react-autosuggest": "^9.4.3",
```
Comment (wasibani-roy) [inline package.json]: Hello @CryceTruly what does this package do??

File: src/components/article/articleComponent/tagComponent.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import React from "react";
+import { Link } from "react-router-dom";
+
+const TagComponent = (props) => {
+  const { tag } = props;
+  return (
+    <Link className="mt-1" to="#">{tag}</Link>
```
Comment (wasibani-roy) [inline src/components/article/articleComponent/tagComponent.js:7]: Do you intend to have the individual tags clickable at some point??

**Review by wasibani-roy — CHANGES_REQUESTED** [review]
Good work @CryceTruly just a few issues that you throw more light on but otherwise well done

File: package.json (hunk 1 above)
Comment (CryceTruly) [reply package.json]: Its a react library that eases displaying a list of suggestions.

**Review by CryceTruly — COMMENTED** [review]
Yes @wasibani-roy, Later we will want to click on a link that displays all articles with a specific tag

**Comment by Peace-Apple** [comment]
Great work @CryceTruly, however, could you add a brief description to your screenshots such that it is easier to know what they are showing when someone looks at them.

**Comment by CryceTruly** [comment]
> Great work @CryceTruly, however, could you add a brief description to your screenshots such that it is easier to know what they are showing when someone looks at them.

Okay @Peace-Apple, thanks for the tip.

File: src/components/article/articleComponent/tagComponent.js (hunk 2 above)
Comment (CryceTruly) [reply src/components/article/articleComponent/tagComponent.js:7]: Yes @wasibani-roy, Later we will want to click on a link that displays all articles with a specific tag

**Review by wasibani-roy — APPROVED** [review]

**Review by BagzieGracious — APPROVED** [review]
Thank you @CryceTruly, it works on properly

**Review by Peace-Apple — APPROVED** [review]
Great work @CryceTruly and thanks for implementing the feedback

File: src/__tests__/components/articleEditor/__test__/articleEditor.test.js [hunk 3]
```diff
@@ -1,11 +1,32 @@
...
+      title: "test titke bero hello test.",
+      body: "test",
+      description: "This is the application",
+      tagList: ["aaa", "hehe", "hi", "hii", "hoho", "huhu", "test", "test."],
+    };
+    localStorage.setItem("article", JSON.stringify(article));
+    const wrapper = shallow(<EditArticleView />);
+    wrapper.instance().handleChange(["one", "ONE", "Test"]);
+    expect(wrapper.instance().state.tagList).toEqual(["one", "test"]);
+  });
```
Comment (AmosWels) [inline src/__tests__/components/articleEditor/__test__/articleEditor.test.js:26]: add a newline here

File: src/components/article/sidebar/index.js [hunk 4]
```diff
@@ -1,7 +1,6 @@
-import React, { Component } from "react";
-
-export default class Sidebar extends Component {
-  render() {
+import React from "react";
+import Tags from "../../../views/tags";
```
Comment (AmosWels) [inline src/components/article/sidebar/index.js:2]: add an extra line here

File: src/views/tags.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import React, { Component } from 'react'
+import {connect} from 'react-redux';
+import {getTags} from "../actions/tags/tags";
+import TagComponent from '../components/article/articleComponent/tagComponent';
+class Tags extends Component {
```
Comment (AmosWels) [inline src/views/tags.js:6]: extra line is required here

File: src/views/tags.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    componentDidMount(){
+       this.props.getTags()
+    }
+  render() {
+      const { isLoading, tags } = this.props.tags;
+    return (
+        <ul className="taglist">
+            {isLoading?"Loading tags":
+                tags.map(tag => (<li><TagComponent tag={tag} key={tag} /></li>))
```
Comment (AmosWels) [inline src/views/tags.js]: Does this cater for an empty tag array? Is there a specific message passed?

File: src/__tests__/reducers/tags/tagsReducer.test.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import {TAGS_LOADED } from '../../../actions/types';
+import tags from '../../../reducers/tags/tags';
+
+  describe("Tags Reducer Tests", () => {
+    const initialState = {
+        tags: [],
+        isLoading: false
+    };
```
Comment (AmosWels) [inline src/__tests__/reducers/tags/tagsReducer.test.js:8]: add a new line here

**Review by AmosWels — CHANGES_REQUESTED** [review]
Goodwork @CryceTruly , please resolve the eslint issues on your PR and respond to the reviews given.

File: src/__tests__/components/articleEditor/__test__/articleEditor.test.js (hunk 3 above)
Comment (CryceTruly) [reply src/__tests__/components/articleEditor/__test__/articleEditor.test.js:26]: I have added a new  line here

File: src/components/article/sidebar/index.js (hunk 4 above)
Comment (CryceTruly) [reply src/components/article/sidebar/index.js:2]: Thanks, @AmosWels, I have added a new line here

File: src/views/tags.js (hunk 5 above)
Comment (CryceTruly) [reply src/views/tags.js:6]: Thanks, @AmosWels, I have added an extra line here

File: src/views/tags.js (hunk 6 above)
Comment (CryceTruly) [reply src/views/tags.js]: No @AmosWels, it actually doesn't but a fix for this has been added.

File: src/__tests__/reducers/tags/tagsReducer.test.js (hunk 7 above)
Comment (CryceTruly) [reply src/__tests__/reducers/tags/tagsReducer.test.js:8]: Thanks, @AmosWels, I have added a new line here

**Review by AmosWels — APPROVED** [review]

**Pull request merged**